probe_src/probe_frontend/src/arena.rs

```diff
@@ -0,0 +1,370 @@
+            true => Some(unsafe { self.raw.as_ptr().add(ptr - self.header.base_address) }),
+        }
+    }
+}
+
+/// this struct represents a `<TID>/data` directory from libprobe.
+pub struct ArenaContext(pub Vec<DataArena>);
+
+impl ArenaContext {
+    pub fn try_deref(&self, ptr: usize) -> Option<*const u8> {
+        for vec in self.0.iter() {
+            if let Some(x) = vec.try_deref(ptr) {
+                return Some(x);
+            }
+        }
+        None
+    }
+}
+
+/// Parse the front of a raw byte buffer into a libprobe arena header
+///
+/// # Safety:
+/// Invoking this function on any byte buffer smaller than [`std::mem::size_of<ArenaHeader>()`]
+/// bytes is undefined behavior (best case a segfault). Invoking this method on a byte buffer
+/// that's not a valid libprobe arena will produce garbage values that should not be used.
+unsafe fn get_header_unchecked(bytes: &[u8]) -> ArenaHeader {
+    let ptr = bytes as *const [u8] as *const ArenaHeader;
+    log::debug!("[unsafe] converting byte buffer into ArenaHeader");
+    unsafe {
+        ArenaHeader {
+            instantiation: (*ptr).instantiation,
+            base_address: (*ptr).base_address,
+            capacity: (*ptr).capacity,
+            used: (*ptr).used,
+        }
+    }
+}
+
+/// Gets the filename from a path and returns it parsed as an integer.
+///
+/// errors if the path has no filename or the filename can't be parsed as an integer.
+fn filename_numeric<P: AsRef<Path>>(dir: P) -> Result<usize> {
+    let filename = dir
+        .as_ref()
+        .file_name()
+        .ok_or_else(|| eyre!("'{}' has no filename", dir.as_ref().to_string_lossy()))?;
+
+    filename
+        .to_str()
+        .ok_or_else(|| eyre!("filename '{}' not valid UTF-8", filename.to_string_lossy()))?
+        .parse::<usize>()
+        .wrap_err(format!(
+            "unable to convert filename '{}' to integer",
+            filename.to_string_lossy()
+        ))
+}
+
+/// Recursively parse a TID libprobe arena allocator directory from `in_dir` and write it in
+/// serialized format to `out_dir`.
+///
+/// This function parses a TID directory in 6 steps:
+///
+/// 1. Output file is created.
+/// 2. Paths of sub-directory are parsed into a [`HashMap`].
+/// 3. `data` directory is is read and parsed into [`DataArena`]s which are then parsed into an
+///    [`ArenaContext`].
+/// 4. `ops` directory is read and parsed into [`OpsArena`]s.
+/// 5. [`OpsArena`]s are parsed into which are then parsed into [`ops::Op`]s using the
+///    [`ArenaContext`].
+/// 6. [`ops::Op`]s are serialized into json and written line-by-line into the output directory.
+///
+/// (steps 5 & 6 are done with iterators to reduce unnecessary memory allocations)
+fn parse_tid<P1: AsRef<Path>, P2: AsRef<Path>>(in_dir: P1, out_dir: P2) -> Result<()> {
+    fn try_files_from_arena_dir<P: AsRef<Path>>(dir: P) -> Result<Vec<PathBuf>> {
+        match fs::read_dir(&dir) {
+            Ok(x) => x
```

Comment:
Let's call the first `x` as `child` and the second `x` as `grandchild` and the third `x` as `great_grandchild`.

Reply:
I did rename them, but not to those names, as those names weren't really representative of what those bindings actually are; I'm so used to having type hints in my editor I forgot that the difference between `map()`-ing over an `impl Iter` and a `Result` is not always obvious.